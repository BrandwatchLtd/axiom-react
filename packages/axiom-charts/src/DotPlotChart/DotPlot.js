import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classnames from 'classnames';
import { Base } from '@brandwatch/axiom-components';
import DotPlotBenchmarkLine from './DotPlotBenchmarkLine';
import DotPlotContext from './DotPlotContext';
import DotPlotDifferenceLine from './DotPlotDifferenceLine';
import DotPlotValue from './DotPlotValue';
import {
  getLines,
  getDotColors,
  isBenchmarkFaded,
  isDotFaded,
  isDotHidden,
  isValueHidden,
  isValueStrong,
} from './utils';
import './DotPlot.css';

const differenceLineContainerClasses = (fromBenchmark, toBenchmark) =>
  classnames('ax-dot-plot__difference-line-container', {
    'ax-dot-plot__difference-line-container--from-benchmark': fromBenchmark,
    'ax-dot-plot__difference-line-container--to-benchmark': toBenchmark,
  });

export default class DotPlot extends Component {
  static propTypes = {
    DropdownContext: PropTypes.func,
    benchmark: PropTypes.number,
    data: PropTypes.arrayOf(PropTypes.shape({
      colors: PropTypes.arrayOf(PropTypes.string).isRequired,
      value: PropTypes.number.isRequired,
    })).isRequired,
    dotPlotLabel: PropTypes.func,
    label: PropTypes.string,
    lower: PropTypes.number,
    mouseOverColors: PropTypes.arrayOf(PropTypes.string),
    mouseOverRowIndex: PropTypes.number,
    onDotMouseEnter: PropTypes.func.isRequired,
    onDotMouseLeave: PropTypes.func.isRequired,
    rawData: PropTypes.object.isRequired,
    rowIndex: PropTypes.number.isRequired,
    upper: PropTypes.number,
  };

  render() {
    const {
      DropdownContext,
      benchmark,
      data,
      dotPlotLabel,
      label,
      lower,
      mouseOverColors,
      mouseOverRowIndex,
      onDotMouseEnter,
      onDotMouseLeave,
      rawData,
      rowIndex,
      upper,
      ...rest
    } = this.props;

    let benchmarkValue;
    if (benchmark) {
      benchmarkValue = ((benchmark - lower) / (upper - lower)) * 100;
    }

    return (
      <Base { ...rest } className="ax-dot-plot">
        { getLines(data, benchmark, mouseOverRowIndex, mouseOverColors, rowIndex, lower, upper)
          .map(({ fromBenchmark, toBenchmark, faded, fromX, toX }) =>
            <div
                className={ differenceLineContainerClasses(fromBenchmark, toBenchmark) }
                key={ `${fromX}.${toX}` }>
              <DotPlotDifferenceLine faded={ faded } from={ fromX } to={ toX } />
            </div>
          )
        }

        { data.map(({ colors, value }) =>
          <DotPlotContext
              DropdownContext={ DropdownContext }
              colors={ getDotColors(mouseOverRowIndex, mouseOverColors, rowIndex, colors) }
              data={ rawData }
              faded={ isDotFaded(mouseOverRowIndex, mouseOverColors, rowIndex, colors) }
              hidden={ isDotHidden(mouseOverRowIndex, mouseOverColors, rowIndex, colors) }
              key={ value }
              label={ label }
              lower={ lower }
              onMouseEnter={ () => onDotMouseEnter(colors) }
              onMouseLeave={ onDotMouseLeave }
              upper={ upper }
              value={ value } />
        ) }

        { benchmarkValue !== undefined && (
          <div className="ax-dot-plot__benchmark-line-container">
            <DotPlotBenchmarkLine
                faded={ isBenchmarkFaded(mouseOverRowIndex, mouseOverColors, rowIndex) }
                value={ benchmarkValue } />
          </div>
        ) }

        { data.map(({ colors, value }) =>
          <DotPlotValue
              dotPlotLabel={ dotPlotLabel }
              hidden={ isValueHidden(mouseOverRowIndex, mouseOverColors, rowIndex, colors) }
              key={ value }
              textStrong={ isValueStrong(mouseOverRowIndex, mouseOverColors, rowIndex, colors) }
              value={ value }
              x={ ((value - lower) / (upper - lower)) * 100 } />
        ) }
      </Base>
    );
  }
}
