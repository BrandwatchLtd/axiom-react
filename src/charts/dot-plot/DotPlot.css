@import '../../materials/animations';
@import '../../materials/colors';
@import '../../materials/typography';
@import '../vars';

:root {
  --cmp-dot-plot-difference-line-width: 0.0625rem;
}

.ax-dot-plot {
  display: block;
  position: relative;
  width: calc(100% - var(--cmp-chart-overflow-space));
  height: var(--cmp-dot-plot-diameter);
  margin-left: calc(var(--cmp-dot-plot-radius) * -1);
}

.ax-dot-plot__dots {
  display: flex;
  position: absolute;
  align-items: center;
  transition-property: opacity;
  transition-duration: var(--transition-time-base);
  transition-timing-function: var(--transition-function);
}

.ax-dot-plot__value {
  display: flex;
  position: absolute;
  top: 50%;
  margin-left: var(--cmp-chart-label-margin);
  transform: translate(var(--cmp-dot-plot-diameter), -50%);
  line-height: var(--cmp-dot-plot-diameter);
  transition-property: opacity;
  transition-duration: var(--transition-time-base);
  transition-timing-function: var(--transition-function);
  pointer-events: none;
}

.ax-dot-plot__benchmark-line-container,
.ax-dot-plot__difference-line-container {
  position: absolute;
  width: 100%;
  height: 100%;
  pointer-events: none;
}

.ax-dot-plot__benchmark-line-container  {
  top: 50%;
}

.ax-dot-plot__benchmark-line,
.ax-dot-plot__difference-line {
  position: relative;
  transition-property: opacity;
  transition-duration: var(--transition-time-base);
  transition-timing-function: var(--transition-function);
}

.ax-dot-plot__benchmark-line-path {
  position: absolute;
  top: 50%;
  height: var(--cmp-benchmark-line-height);
  transform: translate(0, -50%);
  border-right: var(--cmp-benchmark-line-width) solid var(--color-context-text-disabled);
}

.ax-dot-plot__difference-line-path {
  position: absolute;
  top: 50%;
  border-top: var(--cmp-dot-plot-difference-line-width) solid var(--color-context-text-disabled);
}

.ax-dot-plot__difference-line-container--dot .ax-dot-plot__difference-line-path {
  margin-left: var(--cmp-dot-plot-diameter);
}

.ax-dot-plot__dots--hidden,
.ax-dot-plot__difference-line--hidden,
.ax-dot-plot__value--hidden {
  opacity: 0;
}

.ax-dot-plot__benchmark-line--faded,
.ax-dot-plot__difference-line--faded,
.ax-dot-plot__dots--faded {
  opacity: 0.25;
}
