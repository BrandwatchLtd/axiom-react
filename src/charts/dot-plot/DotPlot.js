import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { DataPoints, DataPoint } from 'bw-axiom';
import DotPlotLine from './DotPlotLine';
import './DotPlot.css';

export default class DotPlot extends Component {
  static propTypes = {
    data: PropTypes.arrayOf(PropTypes.shape({
      color: PropTypes.oneOf([
        'rose',
        'pink',
        'purple',
        'lilac',
        'blue',
        'teal',
        'green',
        'chartreuse',
        'amber',
        'orange',
        'brown',
        'grey',
      ]).isRequired,
      value: PropTypes.number.isRequired,
    })).isRequired,
  };

  render() {
    const { data } = this.props;
    const diameter = 0.8125;
    const height = `${diameter}rem`;
    const xValues = data.map(({ value }) => value);
    const x = Math.min(...xValues);
    const range = Math.max(...xValues) - x;
    const style = {
      height,
      marginLeft: `-${diameter / 2}rem`,
    };

    // [{color: String, value: Number}] -> [{colors: [String], value: Number}]
    const stackedData = data.reduce((acc, val) => {
      for (const data of acc) {
        if (data.value === val.value) {
          data.colors.push(val.color);
          return acc;
        }
      }
      acc.push({ value: val.value, colors: [val.color] });
      return acc;
    }, []);

    return (
      <svg
          className="ax-dot-plot"
          style={ style }>
        <DotPlotLine width={ `${ range }%` } x={ `${ x }%` } />
        { stackedData.map(({ value, colors }, i) => (
          <DataPoints
              key={ i }
              preserveAspectRatio="xMinYMin meet"
              size={ height }
              x={ `${value}%` } >
            {colors.map((color, index) => <DataPoint color={ color } key={ index } />)}
          </DataPoints>
        )) }
      </svg>
    );
  }
}
