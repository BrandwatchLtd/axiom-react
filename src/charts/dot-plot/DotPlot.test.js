import React from 'react';
import renderer from 'react-test-renderer';
import DotPlot from './DotPlot';

function getComponent(props) {
  return renderer.create(<DotPlot { ...props } />);
}

describe('DotPlot', () => {
  it('renders with sorted data', () => {
    const component = getComponent({
      data: [
        { color: 'blue', value: 10 },
        { color: 'pink', value: 20 },
      ],
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with unsorted data', () => {
    const component = getComponent({
      data: [
        { color: 'pink', value: 20 },
        { color: 'blue', value: 10 },
      ],
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
