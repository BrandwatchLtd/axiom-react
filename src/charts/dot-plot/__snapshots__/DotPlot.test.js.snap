// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DotPlot renders with sorted data 1`] = `
<svg
  className="ax-dot-plot"
  style={
    Object {
      "height": "0.8125rem",
      "marginLeft": "-0.40625rem",
    }
  }
>
  <svg
    className="ax-dot-plot__line"
    height="1px"
    preserveAspectRatio="none"
    shapeRendering="crispEdges"
    viewBox="0 0 1 1"
    width="10%"
    x="10%"
    y="50%"
  >
    <path
      className="ax-dot-plot__line-path"
      d="M 0 1 H 1"
      strokeWidth="1"
    />
  </svg>
  <svg
    className="ax-data-points"
    onClick={undefined}
    preserveAspectRatio="xMinYMin meet"
    style={
      Object {
        "height": "0.8125rem",
        "width": "0.8125rem",
      }
    }
    viewBox="0 0 24 24"
    x="10%"
  >
    <circle
      className="ax-data-point ax-data-point--blue"
      cx={12}
      cy={12}
      r={12}
    />
  </svg>
  <svg
    className="ax-data-points"
    onClick={undefined}
    preserveAspectRatio="xMinYMin meet"
    style={
      Object {
        "height": "0.8125rem",
        "width": "0.8125rem",
      }
    }
    viewBox="0 0 24 24"
    x="20%"
  >
    <circle
      className="ax-data-point ax-data-point--pink"
      cx={12}
      cy={12}
      r={12}
    />
  </svg>
</svg>
`;

exports[`DotPlot renders with unsorted data 1`] = `
<svg
  className="ax-dot-plot"
  style={
    Object {
      "height": "0.8125rem",
      "marginLeft": "-0.40625rem",
    }
  }
>
  <svg
    className="ax-dot-plot__line"
    height="1px"
    preserveAspectRatio="none"
    shapeRendering="crispEdges"
    viewBox="0 0 1 1"
    width="10%"
    x="10%"
    y="50%"
  >
    <path
      className="ax-dot-plot__line-path"
      d="M 0 1 H 1"
      strokeWidth="1"
    />
  </svg>
  <svg
    className="ax-data-points"
    onClick={undefined}
    preserveAspectRatio="xMinYMin meet"
    style={
      Object {
        "height": "0.8125rem",
        "width": "0.8125rem",
      }
    }
    viewBox="0 0 24 24"
    x="20%"
  >
    <circle
      className="ax-data-point ax-data-point--pink"
      cx={12}
      cy={12}
      r={12}
    />
  </svg>
  <svg
    className="ax-data-points"
    onClick={undefined}
    preserveAspectRatio="xMinYMin meet"
    style={
      Object {
        "height": "0.8125rem",
        "width": "0.8125rem",
      }
    }
    viewBox="0 0 24 24"
    x="10%"
  >
    <circle
      className="ax-data-point ax-data-point--blue"
      cx={12}
      cy={12}
      r={12}
    />
  </svg>
</svg>
`;
